File: dynamic-programming/leetcode-188/solution.cpp
#include <algorithm>
#include <cstdio>
#include <cstring>
#include <vector>
using namespace std;

class Solution {
public:
  int maxProfit(int k, vector<int> &prices) {
    int n = prices.size();
    int current[101][2];
    int next[101][2];

    // Caso onde posso fazer quantas transações quiser
    if (k >= n / 2) {
      int profit = 0;
      for (int i = 1; i < n; i++) {
        if (prices[i] > prices[i - 1]) {
          profit += prices[i] - prices[i - 1];
        }
      }
      return profit;
    }
    // Inicialização da matriz
    memset(current, 0, sizeof(current));
    memset(next, 0, sizeof(next));

    // Programação dinâmica
    for (int i = n - 1; i >= 0; i--) {
      for (int j = 1; j <= k; j++) {
        // Caso onde não tenho ações
        current[j][0] = max(next[j][0],            // Não fazer nada
                            next[j][1] - prices[i] // Comprar a ação
        );
        // Caso eu tenha ações
        current[j][1] = max(next[j][1],                // Não fazer nada
                            next[j - 1][0] + prices[i] // Vender a ação
        );

        next[j][0] = current[j][0];
        next[j][1] = current[j][1];
      }
    }
    return current[k][0];
  }
};

int main(int argc, char *argv[]) {
  // Teste sua solução com novos casos de teste
  Solution solution;

  vector<int> prices = {2, 4, 1};
  printf("%d\n", solution.maxProfit(2, prices));

  prices = {3, 2, 6, 5, 0, 3};
  printf("%d\n", solution.maxProfit(2, prices));

  prices = {3, 3, 5, 0, 0, 3, 1, 4};
  printf("%d\n", solution.maxProfit(2, prices));

  prices = {3, 3, 5, 0, 0, 3, 1, 4};
  printf("%d\n", solution.maxProfit(2, prices));

  return 0;
}

File: dynamic-programming/leetcode-188/solution.py
from typing import List


class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        n = len(prices)
        """ 
        Indices:
            0 - Não tenho uma ação
            1 - Tenho uma ação
        """
        current = [[0, 0] for _ in range(k + 1)]
        _next = [[0, 0] for _ in range(k + 1)]

        for i in range(n - 1, -1, -1):
            for j in range(1, k + 1):
                # Caso eu não tenha a ação
                current[j][0] = max(
                    _next[j][0],  # Não fazer nada
                    _next[j][1] - prices[i],  # Comprar a ação
                )
                # Caso eu tenha a ação
                current[j][1] = max(
                    _next[j][1],  # Não fazer nada
                    _next[j - 1][0] + prices[i],  # Vender a ação
                )

                _next[j][0] = current[j][0]
                _next[j][1] = current[j][1]

        return current[k][0]


def test_solution():
    s = Solution()
    assert s.maxProfit(2, [2, 4, 1]) == 2
    assert s.maxProfit(2, [3, 2, 6, 5, 0, 3]) == 7
    assert s.maxProfit(0, [1, 2, 3, 4, 5]) == 0
    assert s.maxProfit(1, [7, 6, 4, 3, 1]) == 0
    assert s.maxProfit(2, [1, 2, 4, 2, 5, 7, 2, 4, 9, 0]) == 13
